### PR TITLE
fix(async component): memory leak after synchronous async loading (fix #9229)

### LINKED ISSUE #9229: Memory Leak When Using Synchronous Loading With Async Components
### Version
2.5.21

### Reproduction link
[https://codesandbox.io/s/xlly27wy8z](https://codesandbox.io/s/xlly27wy8z)

### Steps to reproduce
Same as: https://github.com/vuejs/vue/issues/8740

- Load the reproduction link in Chrome
- Open Chrome Dev Tools
- Go to the memory tab
- Click on the "Destroy" button
- Take a "Heap Snapshot"
- Filter to look for "Detached HTMLDivElement"
- You should see 2 leaking HTMLDivElement

### What is expected?
Context for loading the Async component should be cleared.

### What is actually happening?
There are 2 components still attached to a context of an Async loading Component.
It is not cleared if the Async component is loaded synchronously.

---
This is the same issue as https://github.com/vuejs/vue/issues/8740. However the fix for it only fixes the case when loading is actually async. Synchronous async loading does not call the "forceRender" (https://github.com/vuejs/vue/commit/2e472c5e5e559a7a4083b4164ffe0c3911ce0651#diff-293f31f6a318619ad803b971ef0724a3R64) so the contexts array is not cleared.
I can try to fix it and submit a PR if needed.

### PR DESCRIPTION
**Pull request by adrienbaron**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

File: test/unit/modules/vdom/create-component.spec.js [hunk 1]
```diff
@@ -76,6 +76,32 @@ describe('create-component', () => {
...
+    expect(vnode.asyncFactory).toBe(async)
+    expect(vnode.asyncFactory.contexts.length).toEqual(0)
+    expect(vnode.tag).toMatch(/vue-component-[0-9]+-child/)
+    expect(vnode.data.staticAttrs).toEqual({ class: 'bar' })
+    expect(vnode.children).toBeUndefined()
+    expect(vnode.text).toBeUndefined()
+    expect(vnode.elm).toBeUndefined()
+    expect(vnode.ns).toBeUndefined()
+    expect(vnode.context).toEqual(vm)
+     expect(vm.$forceUpdate).not.toHaveBeenCalled()
```
Comment (phanan) [inline test/unit/modules/vdom/create-component.spec.js]: Did my suggested patch include an extra space 😅? Sorry for that.

Comment (adrienbaron) [inline test/unit/modules/vdom/create-component.spec.js]:
```suggestion
    expect(vm.$forceUpdate).not.toHaveBeenCalled()
```

Comment (adrienbaron) [reply test/unit/modules/vdom/create-component.spec.js]: Apparently ^^, just fixed it 👍

**Comment by adrienbaron** [comment]
Hi! Any news on the progress of this PR validation?

**Pull request merged**